File: app/src/main/java/com/fiatlife/app/domain/model/Bill.kt
package com.fiatlife.app.domain.model

import kotlinx.serialization.Serializable

@Serializable
data class BillPayment(
    val date: Long = 0L,
    val amount: Double = 0.0
)

@Serializable
data class StatementEntry(
    val hash: String = "",
    val addedAt: Long = 0L,
    val label: String = ""
)

/** How the credit card minimum payment is determined. */
@Serializable
enum class CreditCardMinPaymentType {
    /** Fixed dollar amount (e.g. $25). */
    FIXED,
    /** Percentage of current balance (e.g. 2%). */
    PERCENT_OF_BALANCE,
    /** Pay full balance (minimum = balance). */
    FULL_BALANCE
}

@Serializable
data class CreditCardDetails(
    /** Current balance owed (updated when you record a payment or enter a new statement). */
    val currentBalance: Double = 0.0,
    /** Annual percentage rate (e.g. 0.1999 for 19.99% APR). */
    val apr: Double = 0.0,
    val minimumPaymentType: CreditCardMinPaymentType = CreditCardMinPaymentType.PERCENT_OF_BALANCE,
    /** For FIXED: dollar amount. For PERCENT_OF_BALANCE: percentage 0–100. Ignored for FULL_BALANCE. */
    val minimumPaymentValue: Double = 2.0,
    /** Optional: interest charged last period (for tracking). */
    val interestChargedLastPeriod: Double = 0.0
) {
    /** Compute minimum payment due based on current balance. */
    fun minimumDue(balance: Double): Double = when (minimumPaymentType) {
        CreditCardMinPaymentType.FIXED -> minimumPaymentValue.coerceAtLeast(0.0)
        CreditCardMinPaymentType.PERCENT_OF_BALANCE -> (balance * (minimumPaymentValue / 100.0)).coerceAtLeast(0.0)
        CreditCardMinPaymentType.FULL_BALANCE -> balance.coerceAtLeast(0.0)
    }

    /** Estimated interest next month: balance * APR/12. */
    fun estimatedMonthlyInterest(balance: Double): Double =
        if (apr <= 0 || balance <= 0) 0.0 else balance * (apr / 12.0)
}

/** General bill category for high-level grouping (header totals, dashboard). */
@Serializable
enum class BillGeneralCategory {
    AUTO,
    UTILITIES,
    HOME,
    HEALTH,
    CREDIT_LOANS,
    SUBSCRIPTION,
    PERSONAL,
    OTHER;

    val displayName: String
        get() = when (this) {
            AUTO -> "Auto"
            UTILITIES -> "Utilities"
            HOME -> "Home"
            HEALTH -> "Health"
            CREDIT_LOANS -> "Credit/Loans"
            SUBSCRIPTION -> "Subscription"
            PERSONAL -> "Personal"
            OTHER -> "Other"
        }
}

/** Specific subcategory under a general category; each has a [generalCategory]. */
@Serializable
enum class BillSubcategory(val generalCategory: BillGeneralCategory) {
    // Home
    MORTGAGE_RENT(BillGeneralCategory.HOME),
    HOME_INSURANCE(BillGeneralCategory.HOME),
    INTERNET(BillGeneralCategory.HOME),
    HOA(BillGeneralCategory.HOME),
    PROPERTY_TAX(BillGeneralCategory.HOME),
    // Utilities
    ELECTRIC(BillGeneralCategory.UTILITIES),
    GAS_HEATING(BillGeneralCategory.UTILITIES),
    WATER_SEWER(BillGeneralCategory.UTILITIES),
    TRASH(BillGeneralCategory.UTILITIES),
    // Auto
    CAR_PAYMENT(BillGeneralCategory.AUTO),
    CAR_INSURANCE(BillGeneralCategory.AUTO),
    GAS_FUEL(BillGeneralCategory.AUTO),
    // Credit/Loans
    CREDIT_CARD(BillGeneralCategory.CREDIT_LOANS),
    STUDENT_LOAN(BillGeneralCategory.CREDIT_LOANS),
    OTHER_LOAN(BillGeneralCategory.CREDIT_LOANS),
    // Subscription
    EDUCATION(BillGeneralCategory.SUBSCRIPTION),
    FINANCE(BillGeneralCategory.SUBSCRIPTION),
    FOOD(BillGeneralCategory.SUBSCRIPTION),
    GAMING(BillGeneralCategory.SUBSCRIPTION),
    HEALTH_WELLNESS(BillGeneralCategory.SUBSCRIPTION),
    SUB_HOME(BillGeneralCategory.SUBSCRIPTION),
    MUSIC(BillGeneralCategory.SUBSCRIPTION),
    NEWS_MEDIA(BillGeneralCategory.SUBSCRIPTION),
    PET_CARE(BillGeneralCategory.SUBSCRIPTION),
    TRAVEL(BillGeneralCategory.SUBSCRIPTION),
    SHOPPING(BillGeneralCategory.SUBSCRIPTION),
    SOFTWARE(BillGeneralCategory.SUBSCRIPTION),
    STREAMING(BillGeneralCategory.SUBSCRIPTION),
    VEHICLE(BillGeneralCategory.SUBSCRIPTION),
    OTHER_SUBSCRIPTION(BillGeneralCategory.SUBSCRIPTION),
    // Health
    GYM_FITNESS(BillGeneralCategory.HEALTH),
    MEDICAL(BillGeneralCategory.HEALTH),
    // Personal
    GROCERIES(BillGeneralCategory.PERSONAL),
    CHILDCARE(BillGeneralCategory.PERSONAL),
    PET(BillGeneralCategory.PERSONAL),
    // Other
    OTHER(BillGeneralCategory.OTHER);

    val displayName: String
        get() = when (this) {
            MORTGAGE_RENT -> "Mortgage/Rent"
            HOME_INSURANCE -> "Home Insurance"
            INTERNET -> "Internet"
            HOA -> "HOA"
            PROPERTY_TAX -> "Property Tax"
            ELECTRIC -> "Electric"
            GAS_HEATING -> "Gas/Heating"
            WATER_SEWER -> "Water/Sewer"
            TRASH -> "Trash"
            CAR_PAYMENT -> "Car Payment"
            CAR_INSURANCE -> "Car Insurance"
            GAS_FUEL -> "Gas/Fuel"
            CREDIT_CARD -> "Credit Card"
            STUDENT_LOAN -> "Student Loan"
            OTHER_LOAN -> "Other Loan"
            EDUCATION -> "Education"
            FINANCE -> "Finance"
            FOOD -> "Food"
            GAMING -> "Gaming"
            HEALTH_WELLNESS -> "Health/Wellness"
            SUB_HOME -> "Home"
            MUSIC -> "Music"
            NEWS_MEDIA -> "News/Media"
            PET_CARE -> "Pet Care"
            TRAVEL -> "Travel"
            SHOPPING -> "Shopping"
            SOFTWARE -> "Software"
            STREAMING -> "Streaming"
            VEHICLE -> "Vehicle"
            OTHER_SUBSCRIPTION -> "Other Subscription"
            GYM_FITNESS -> "Gym/Fitness"
            MEDICAL -> "Medical"
            GROCERIES -> "Groceries"
            CHILDCARE -> "Childcare"
            PET -> "Pet"
            OTHER -> "Other"
        }

    val icon: String
        get() = when (this) {
            MORTGAGE_RENT -> "home"
            HOME_INSURANCE -> "shield"
            INTERNET -> "wifi"
            HOA -> "apartment"
            PROPERTY_TAX -> "account_balance"
            ELECTRIC -> "bolt"
            GAS_HEATING -> "local_fire_department"
            WATER_SEWER -> "water_drop"
            TRASH -> "delete"
            CAR_PAYMENT -> "directions_car"
            CAR_INSURANCE -> "shield"
            GAS_FUEL -> "local_gas_station"
            CREDIT_CARD -> "credit_card"
            STUDENT_LOAN -> "school"
            OTHER_LOAN -> "account_balance"
            EDUCATION -> "school"
            FINANCE -> "account_balance"
            FOOD -> "restaurant"
            GAMING -> "sports_esports"
            HEALTH_WELLNESS -> "fitness_center"
            SUB_HOME -> "home"
            MUSIC -> "music_note"
            NEWS_MEDIA -> "newspaper"
            PET_CARE -> "pets"
            TRAVEL -> "flight"
            SHOPPING -> "shopping_cart"
            SOFTWARE -> "computer"
            STREAMING -> "tv"
            VEHICLE -> "directions_car"
            OTHER_SUBSCRIPTION -> "subscriptions"
            GYM_FITNESS -> "fitness_center"
            MEDICAL -> "medical_services"
            GROCERIES -> "shopping_cart"
            CHILDCARE -> "child_care"
            PET -> "pets"
            OTHER -> "receipt"
        }
}

@Serializable
data class Bill(
    val id: String = "",
    val name: String = "",
    val amount: Double = 0.0,
    @Deprecated("Use subcategory instead") val category: BillCategory = BillCategory.OTHER,
    val subcategory: BillSubcategory? = null,
    val frequency: BillFrequency = BillFrequency.MONTHLY,
    val dueDay: Int = 1,
    val autoPay: Boolean = false,
    val accountName: String = "",
    val notes: String = "",
    val attachmentHashes: List<String> = emptyList(),
    val statementEntries: List<StatementEntry> = emptyList(),
    val paymentHistory: List<BillPayment> = emptyList(),
    val isPaid: Boolean = false,
    val lastPaidDate: Long? = null,
    val createdAt: Long = 0L,
    val updatedAt: Long = 0L,
    val creditCardDetails: CreditCardDetails? = null,
    /** When set, this bill represents the monthly payment for the linked credit/loan account (Debt tab). */
    val linkedCreditAccountId: String? = null
) {
    /** Resolved subcategory: [subcategory] if set, else derived from legacy [category]. */
    val effectiveSubcategory: BillSubcategory
        get() = subcategory ?: fromLegacyCategory(category)

    /** General category for this bill (from [effectiveSubcategory]). */
    val effectiveGeneralCategory: BillGeneralCategory
        get() = effectiveSubcategory.generalCategory

    /** True if this bill is tracked as a credit card (has credit card–specific fields). */
    fun isCreditCard(): Boolean = creditCardDetails != null

    /** True if this bill is a credit card or linked to a credit/loan account (Debt tab). */
    fun isCreditOrLoan(): Boolean = isCreditCard() || linkedCreditAccountId != null

    /** Amount due this period: for credit cards, computed minimum due; otherwise bill.amount. */
    fun effectiveAmountDue(): Double = creditCardDetails?.let { cc ->
        cc.minimumDue(cc.currentBalance)
    } ?: amount

    /** Estimated interest next month (credit cards only). */
    fun estimatedMonthlyInterest(): Double = creditCardDetails?.estimatedMonthlyInterest(creditCardDetails.currentBalance) ?: 0.0

    /** Statements to show: statementEntries if present, else legacy attachmentHashes with placeholder dates. */
    fun statementsOrderedByDate(): List<StatementEntry> {
        if (statementEntries.isNotEmpty()) return statementEntries.sortedByDescending { it.addedAt }
        return attachmentHashes.map { StatementEntry(hash = it, addedAt = updatedAt, label = "Statement") }
            .sortedByDescending { it.addedAt }
    }

    /** Total amount paid in the current calendar year. */
    fun annualTotalPaidSoFar(): Double {
        val cal = java.util.Calendar.getInstance()
        cal.set(java.util.Calendar.MONTH, java.util.Calendar.JANUARY)
        cal.set(java.util.Calendar.DAY_OF_MONTH, 1)
        cal.set(java.util.Calendar.HOUR_OF_DAY, 0)
        cal.set(java.util.Calendar.MINUTE, 0)
        cal.set(java.util.Calendar.SECOND, 0)
        cal.set(java.util.Calendar.MILLISECOND, 0)
        val yearStart = cal.timeInMillis
        return paymentHistory.filter { it.date >= yearStart }.sumOf { it.amount }
    }

    /** Next due date (start of day) in ms. Handles MONTHLY; other frequencies use similar logic. */
    fun nextDueDateMillis(): Long? {
        val cal = java.util.Calendar.getInstance()
        val day = dueDay.coerceIn(1, 31)
        when (frequency) {
            BillFrequency.MONTHLY -> {
                cal.set(java.util.Calendar.DAY_OF_MONTH, day.coerceIn(1, cal.getActualMaximum(java.util.Calendar.DAY_OF_MONTH)))
                cal.set(java.util.Calendar.HOUR_OF_DAY, 0)
                cal.set(java.util.Calendar.MINUTE, 0)
                cal.set(java.util.Calendar.SECOND, 0)
                cal.set(java.util.Calendar.MILLISECOND, 0)
                if (cal.timeInMillis <= System.currentTimeMillis())
                    cal.add(java.util.Calendar.MONTH, 1)
                cal.set(java.util.Calendar.DAY_OF_MONTH, day.coerceIn(1, cal.getActualMaximum(java.util.Calendar.DAY_OF_MONTH)))
                return cal.timeInMillis
            }
            BillFrequency.ANNUALLY -> {
                cal.set(java.util.Calendar.MONTH, java.util.Calendar.JANUARY)
                cal.set(java.util.Calendar.DAY_OF_MONTH, day.coerceIn(1, 31))
                cal.set(java.util.Calendar.HOUR_OF_DAY, 0)
                cal.set(java.util.Calendar.MINUTE, 0)
                cal.set(java.util.Calendar.SECOND, 0)
                cal.set(java.util.Calendar.MILLISECOND, 0)
                if (cal.timeInMillis <= System.currentTimeMillis())
                    cal.add(java.util.Calendar.YEAR, 1)
                return cal.timeInMillis
            }
            else -> {
                cal.set(java.util.Calendar.DAY_OF_MONTH, day.coerceIn(1, cal.getActualMaximum(java.util.Calendar.DAY_OF_MONTH)))
                cal.set(java.util.Calendar.HOUR_OF_DAY, 0)
                cal.set(java.util.Calendar.MINUTE, 0)
                cal.set(java.util.Calendar.SECOND, 0)
                cal.set(java.util.Calendar.MILLISECOND, 0)
                while (cal.timeInMillis <= System.currentTimeMillis()) {
                    when (frequency) {
                        BillFrequency.WEEKLY -> cal.add(java.util.Calendar.WEEK_OF_YEAR, 1)
                        BillFrequency.BIWEEKLY -> cal.add(java.util.Calendar.WEEK_OF_YEAR, 2)
                        BillFrequency.BIMONTHLY -> cal.add(java.util.Calendar.MONTH, 2)
                        BillFrequency.QUARTERLY -> cal.add(java.util.Calendar.MONTH, 3)
                        BillFrequency.SEMIANNUALLY -> cal.add(java.util.Calendar.MONTH, 6)
                        else -> cal.add(java.util.Calendar.MONTH, 1)
                    }
                    cal.set(java.util.Calendar.DAY_OF_MONTH, day.coerceIn(1, cal.getActualMaximum(java.util.Calendar.DAY_OF_MONTH)))
                }
                return cal.timeInMillis
            }
        }
    }

    /** Most recent due date (start of day) that is <= now; null if none. Used to detect past due. */
    fun lastDueDateMillis(): Long? {
        val next = nextDueDateMillis() ?: return null
        val cal = java.util.Calendar.getInstance()
        cal.timeInMillis = next
        when (frequency) {
            BillFrequency.MONTHLY -> cal.add(java.util.Calendar.MONTH, -1)
            BillFrequency.WEEKLY -> cal.add(java.util.Calendar.WEEK_OF_YEAR, -1)
            BillFrequency.BIWEEKLY -> cal.add(java.util.Calendar.WEEK_OF_YEAR, -2)
            BillFrequency.BIMONTHLY -> cal.add(java.util.Calendar.MONTH, -2)
            BillFrequency.QUARTERLY -> cal.add(java.util.Calendar.MONTH, -3)
            BillFrequency.SEMIANNUALLY -> cal.add(java.util.Calendar.MONTH, -6)
            BillFrequency.ANNUALLY -> cal.add(java.util.Calendar.YEAR, -1)
        }
        val lastDue = cal.timeInMillis
        return if (lastDue <= System.currentTimeMillis()) lastDue else null
    }

    /** True if this bill is not paid and the due date (end of due day) has passed. */
    fun isPastDue(): Boolean {
        if (isPaid) return false
        val lastDue = lastDueDateMillis() ?: return false
        val endOfDueDay = lastDue + 86400_000L - 1
        return System.currentTimeMillis() > endOfDueDay
    }
}

private fun fromLegacyCategory(category: BillCategory): BillSubcategory = when (category) {
    BillCategory.MORTGAGE_RENT -> BillSubcategory.MORTGAGE_RENT
    BillCategory.ELECTRIC -> BillSubcategory.ELECTRIC
    BillCategory.GAS_HEATING -> BillSubcategory.GAS_HEATING
    BillCategory.WATER_SEWER -> BillSubcategory.WATER_SEWER
    BillCategory.TRASH -> BillSubcategory.TRASH
    BillCategory.INTERNET -> BillSubcategory.INTERNET
    BillCategory.PHONE -> BillSubcategory.OTHER_SUBSCRIPTION
    BillCategory.CABLE_STREAMING -> BillSubcategory.STREAMING
    BillCategory.CAR_PAYMENT -> BillSubcategory.CAR_PAYMENT
    BillCategory.CAR_INSURANCE -> BillSubcategory.CAR_INSURANCE
    BillCategory.HOME_INSURANCE -> BillSubcategory.HOME_INSURANCE
    BillCategory.PROPERTY_TAX -> BillSubcategory.PROPERTY_TAX
    BillCategory.HOA -> BillSubcategory.HOA
    BillCategory.GROCERIES -> BillSubcategory.GROCERIES
    BillCategory.GAS_FUEL -> BillSubcategory.GAS_FUEL
    BillCategory.CHILDCARE -> BillSubcategory.CHILDCARE
    BillCategory.STUDENT_LOAN -> BillSubcategory.STUDENT_LOAN
    BillCategory.CREDIT_CARD -> BillSubcategory.CREDIT_CARD
    BillCategory.SUBSCRIPTION -> BillSubcategory.OTHER_SUBSCRIPTION
    BillCategory.GYM_FITNESS -> BillSubcategory.GYM_FITNESS
    BillCategory.PET -> BillSubcategory.PET
    BillCategory.OTHER -> BillSubcategory.OTHER
}

@Serializable
enum class BillCategory {
    MORTGAGE_RENT,
    ELECTRIC,
    GAS_HEATING,
    WATER_SEWER,
    TRASH,
    INTERNET,
    PHONE,
    CABLE_STREAMING,
    CAR_PAYMENT,
    CAR_INSURANCE,
    HOME_INSURANCE,
    PROPERTY_TAX,
    HOA,
    GROCERIES,
    GAS_FUEL,
    CHILDCARE,
    STUDENT_LOAN,
    CREDIT_CARD,
    SUBSCRIPTION,
    GYM_FITNESS,
    PET,
    OTHER;

    val displayName: String
        get() = when (this) {
            MORTGAGE_RENT -> "Mortgage/Rent"
            ELECTRIC -> "Electric"
            GAS_HEATING -> "Gas/Heating"
            WATER_SEWER -> "Water/Sewer"
            TRASH -> "Trash"
            INTERNET -> "Internet"
            PHONE -> "Phone"
            CABLE_STREAMING -> "Cable/Streaming"
            CAR_PAYMENT -> "Car Payment"
            CAR_INSURANCE -> "Car Insurance"
            HOME_INSURANCE -> "Home Insurance"
            PROPERTY_TAX -> "Property Tax"
            HOA -> "HOA"
            GROCERIES -> "Groceries"
            GAS_FUEL -> "Gas/Fuel"
            CHILDCARE -> "Childcare"
            STUDENT_LOAN -> "Student Loan"
            CREDIT_CARD -> "Credit Card"
            SUBSCRIPTION -> "Subscription"
            GYM_FITNESS -> "Gym/Fitness"
            PET -> "Pet"
            OTHER -> "Other"
        }

    val icon: String
        get() = when (this) {
            MORTGAGE_RENT -> "home"
            ELECTRIC -> "bolt"
            GAS_HEATING -> "local_fire_department"
            WATER_SEWER -> "water_drop"
            TRASH -> "delete"
            INTERNET -> "wifi"
            PHONE -> "phone_android"
            CABLE_STREAMING -> "tv"
            CAR_PAYMENT -> "directions_car"
            CAR_INSURANCE -> "shield"
            HOME_INSURANCE -> "security"
            PROPERTY_TAX -> "account_balance"
            HOA -> "apartment"
            GROCERIES -> "shopping_cart"
            GAS_FUEL -> "local_gas_station"
            CHILDCARE -> "child_care"
            STUDENT_LOAN -> "school"
            CREDIT_CARD -> "credit_card"
            SUBSCRIPTION -> "subscriptions"
            GYM_FITNESS -> "fitness_center"
            PET -> "pets"
            OTHER -> "receipt"
        }
}

@Serializable
enum class BillFrequency(val timesPerYear: Int) {
    WEEKLY(52),
    BIWEEKLY(26),
    MONTHLY(12),
    BIMONTHLY(6),
    QUARTERLY(4),
    SEMIANNUALLY(2),
    ANNUALLY(1);

    val displayName: String
        get() = name.lowercase().replaceFirstChar { it.uppercase() }
}

File: apps/web/web/src/lib/bill.ts
/** Bill domain types aligned with Android `domain/model/Bill.kt`. */

import {
  fromLegacyCategory,
  SUBCATEGORY_GENERAL,
  type BillSubcategory,
} from "./billSubcategory";
export {
  fromLegacyCategory,
  generalForSubcategory,
  subcategoriesForGeneral,
  subcategoryLabel,
  SUBCATEGORY_LABELS,
  type BillSubcategory,
} from "./billSubcategory";
import {
  dueAmountInMonth,
  effectiveAmountDue,
} from "./billSchedule";

export {
  daysUntilDue,
  dueAmountInMonth,
  dueAmountInYear,
  dueOccurrencesInMonth,
  effectiveAmountDue,
  formatDueCountdown,
  isCreditCard,
  isCreditOrLoan,
  isPaidForCurrentCycle,
  isPastDue,
  lastDueDateMillis,
  nextDueDateMillis,
  skippedNextDueDateMillis,
} from "./billSchedule";

export type BillFrequency =
  | "WEEKLY"
  | "BIWEEKLY"
  | "MONTHLY"
  | "BIMONTHLY"
  | "QUARTERLY"
  | "SEMIANNUALLY"
  | "ANNUALLY";

export type BillRecurrenceUnit = "DAY" | "WEEK" | "MONTH" | "YEAR";

export type BillGeneralCategory =
  | "AUTO"
  | "UTILITIES"
  | "HOME"
  | "HEALTH"
  | "CREDIT_LOANS"
  | "SUBSCRIPTION"
  | "PERSONAL"
  | "OTHER";

export type BillSource = "NATIVE" | "CYPHERLOG";

export type BillPayment = {
  date: number;
  amount: number;
};

export type StatementEntry = {
  hash: string;
  addedAt: number;
  label: string;
};

export type BillStatusEvent = {
  date: number;
  type: string;
  note: string;
};

export type CreditCardMinPaymentType =
  | "FIXED"
  | "PERCENT_OF_BALANCE"
  | "FULL_BALANCE";

export type CreditCardDetails = {
  currentBalance?: number;
  apr?: number;
  minimumPaymentType?: CreditCardMinPaymentType;
  minimumPaymentValue?: number;
  interestChargedLastPeriod?: number;
};

export type Bill = {
  id: string;
  name: string;
  amount: number;
  subcategory?: string | null;
  category?: string;
  frequency: BillFrequency;
  dueDay?: number;
  autoPay?: boolean;
  renewalDateMillis?: number | null;
  initialPurchaseDateMillis?: number | null;
  recurrenceUnit?: BillRecurrenceUnit | null;
  recurrenceIntervalCount?: number;
  recurrenceTimezone?: string | null;
  isRecurring?: boolean;
  rateValidUntilMillis?: number | null;
  isCancelled?: boolean;
  cancelledAt?: number | null;
  statusHistory?: BillStatusEvent[];
  accountName?: string;
  notes?: string;
  attachmentHashes?: string[];
  statementEntries?: StatementEntry[];
  paymentHistory?: BillPayment[];
  isPaid?: boolean;
  lastPaidDate?: number | null;
  createdAt?: number;
  updatedAt?: number;
  creditCardDetails?: CreditCardDetails | null;
  linkedCreditAccountId?: string | null;
  linkedBillerId?: string | null;
  billerName?: string;
  payFromBankAccountId?: string | null;
  payFromCreditAccountId?: string | null;
};

export type BillWithSource = {
  bill: Bill;
  source: BillSource;
  dTag: string;
  isCypherLog?: boolean;
  /** Unmapped CypherLog tags preserved for round-trip publish. */
  preservedTags?: Record<string, string[]>;
};

export const GENERAL_CATEGORY_LABELS: Record<BillGeneralCategory, string> = {
  AUTO: "Auto",
  UTILITIES: "Utilities",
  HOME: "Home",
  HEALTH: "Health",
  CREDIT_LOANS: "Credit/Loans",
  SUBSCRIPTION: "Subscription",
  PERSONAL: "Personal",
  OTHER: "Other",
};

export const ALL_GENERAL_CATEGORIES: BillGeneralCategory[] = [
  "HOME",
  "UTILITIES",
  "AUTO",
  "CREDIT_LOANS",
  "SUBSCRIPTION",
  "HEALTH",
  "PERSONAL",
  "OTHER",
];

export function effectiveSubcategory(bill: Bill): BillSubcategory {
  if (bill.subcategory) {
    return bill.subcategory as BillSubcategory;
  }
  if (bill.category) return fromLegacyCategory(bill.category);
  return "OTHER";
}

export function generalCategoryForBill(bill: Bill): BillGeneralCategory {
  const sub = effectiveSubcategory(bill);
  return SUBCATEGORY_GENERAL[sub] ?? "OTHER";
}

export function monthlyEquivalent(bill: Bill, monthAnchor = Date.now()): number {
  if (bill.isCancelled) return 0;
  if (bill.isRecurring === false) return 0;
  return dueAmountInMonth(bill, monthAnchor) || 0;
}

export function frequencyLabel(freq: BillFrequency): string {
  return freq.charAt(0) + freq.slice(1).toLowerCase().replace(/_/g, " ");
}

export function canSkipInterval(bill: Bill): boolean {
  return (
    bill.isRecurring !== false &&
    !bill.isCancelled &&
    generalCategoryForBill(bill) === "SUBSCRIPTION" &&
    (effectiveSubcategory(bill) === "FOOD" ||
      effectiveSubcategory(bill) === "HEALTH_WELLNESS")
  );
}

export function statementsOrderedByDate(bill: Bill): StatementEntry[] {
  const entries = bill.statementEntries ?? [];
  if (entries.length > 0) {
    return [...entries].sort((a, b) => b.addedAt - a.addedAt);
  }
  const hashes = bill.attachmentHashes ?? [];
  const updated = bill.updatedAt ?? 0;
  return hashes
    .map((hash) => ({ hash, addedAt: updated, label: "Statement" }))
    .sort((a, b) => b.addedAt - a.addedAt);
}

export function annualTotalPaidSoFar(bill: Bill, now = Date.now()): number {
  const d = new Date(now);
  d.setMonth(0);
  d.setDate(1);
  d.setHours(0, 0, 0, 0);
  const yearStart = d.getTime();
  return (bill.paymentHistory ?? [])
    .filter((p) => p.date >= yearStart)
    .reduce((sum, p) => sum + p.amount, 0);
}

function startOfDay(ms: number): number {
  const d = new Date(ms);
  d.setHours(0, 0, 0, 0);
  return d.getTime();
}

export function parseBillRecord(
  dTag: string,
  plaintext: string,
): BillWithSource | null {
  try {
    const parsed = JSON.parse(plaintext) as Record<string, unknown>;
    if (parsed.deleted === true) return null;

    const source: BillSource = dTag.startsWith("subscription:")
      ? "CYPHERLOG"
      : "NATIVE";

    const bill: Bill = {
      id: String(parsed.id ?? dTag.split("/").pop() ?? crypto.randomUUID()),
      name: String(parsed.name ?? "Unnamed bill"),
      amount: Number(parsed.amount ?? 0),
      subcategory: parsed.subcategory != null ? String(parsed.subcategory) : null,
      category: parsed.category != null ? String(parsed.category) : undefined,
      frequency: (parsed.frequency as BillFrequency) ?? "MONTHLY",
      dueDay: Number(parsed.dueDay ?? 1),
      autoPay: Boolean(parsed.autoPay),
      renewalDateMillis:
        parsed.renewalDateMillis != null
          ? Number(parsed.renewalDateMillis)
          : null,
      initialPurchaseDateMillis:
        parsed.initialPurchaseDateMillis != null
          ? Number(parsed.initialPurchaseDateMillis)
          : null,
      recurrenceUnit:
        parsed.recurrenceUnit != null
          ? (String(parsed.recurrenceUnit) as BillRecurrenceUnit)
          : null,
      recurrenceIntervalCount: Number(parsed.recurrenceIntervalCount ?? 1),
      recurrenceTimezone:
        parsed.recurrenceTimezone != null
          ? String(parsed.recurrenceTimezone)
          : null,
      isRecurring: parsed.isRecurring !== false,
      rateValidUntilMillis:
        parsed.rateValidUntilMillis != null
          ? Number(parsed.rateValidUntilMillis)
          : null,
      isCancelled: Boolean(parsed.isCancelled),
      cancelledAt:
        parsed.cancelledAt != null ? Number(parsed.cancelledAt) : null,
      statusHistory: Array.isArray(parsed.statusHistory)
        ? (parsed.statusHistory as BillStatusEvent[])
        : [],
      accountName: parsed.accountName != null ? String(parsed.accountName) : "",
      notes: parsed.notes != null ? String(parsed.notes) : "",
      attachmentHashes: Array.isArray(parsed.attachmentHashes)
        ? (parsed.attachmentHashes as string[])
        : [],
      statementEntries: Array.isArray(parsed.statementEntries)
        ? (parsed.statementEntries as StatementEntry[])
        : [],
      isPaid: Boolean(parsed.isPaid),
      lastPaidDate:
        parsed.lastPaidDate != null ? Number(parsed.lastPaidDate) : null,
      paymentHistory: Array.isArray(parsed.paymentHistory)
        ? (parsed.paymentHistory as BillPayment[])
        : [],
      creditCardDetails:
        parsed.creditCardDetails != null
          ? (parsed.creditCardDetails as CreditCardDetails)
          : null,
      linkedCreditAccountId:
        parsed.linkedCreditAccountId != null
          ? String(parsed.linkedCreditAccountId)
          : null,
      linkedBillerId:
        parsed.linkedBillerId != null ? String(parsed.linkedBillerId) : null,
      billerName: parsed.billerName != null ? String(parsed.billerName) : "",
      payFromBankAccountId:
        parsed.payFromBankAccountId != null
          ? String(parsed.payFromBankAccountId)
          : null,
      payFromCreditAccountId:
        parsed.payFromCreditAccountId != null
          ? String(parsed.payFromCreditAccountId)
          : null,
      createdAt: Number(parsed.createdAt ?? 0),
      updatedAt: Number(parsed.updatedAt ?? 0),
    };

    if (!bill.name.trim()) return null;
    return {
      bill,
      source,
      dTag,
      isCypherLog: source === "CYPHERLOG",
    };
  } catch {
    return null;
  }
}

export function billDTag(id: string, source: BillSource): string {
  return source === "CYPHERLOG" ? `subscription:${id}` : `fiatlife/bill/${id}`;
}

export function newBillId(): string {
  return crypto.randomUUID();
}

export function markBillPaid(
  bill: Bill,
  amount?: number,
  now = Date.now(),
): Bill {
  const paidAmount = amount ?? effectiveAmountDue(bill);
  const history = [...(bill.paymentHistory ?? [])];
  history.push({ date: startOfDay(now), amount: paidAmount });
  return {
    ...bill,
    isPaid: true,
    lastPaidDate: now,
    paymentHistory: history,
    updatedAt: now,
  };
}

export function markBillUnpaid(bill: Bill, now = Date.now()): Bill {
  return {
    ...bill,
    isPaid: false,
    updatedAt: now,
  };
}

export function serializeBill(bill: Bill): string {
  return JSON.stringify(bill);
}
